samples/async_op_samples_int/ml_async_op_module_with_retries.c

```diff
@@ -0,0 +1,458 @@
+                (void)interlocked_exchange(&async_op_context->is_canceled, 0);
+
+                // 4. Initialize the ll_async_op to NULL
+                THANDLE_INITIALIZE(ASYNC_OP)(&async_op_context->ll_async_op, NULL);
+
+                // 5. Store the async_op from the LL in a temporary variable
+                THANDLE(ASYNC_OP) ll_async_op = NULL;
+
+                // 6. Take an additional reference on the async_op, we have 1 reference for returning to the caller (async_op), and 1 reference for the async work callback (async_op_ref_for_callback)
+                //    Need to take the reference before starting the operation below because its callback may come immediately
+                THANDLE(ASYNC_OP) async_op_ref_for_callback = NULL;
+                THANDLE_ASSIGN(ASYNC_OP)(&async_op_ref_for_callback, async_op);
+
+                // 7. Start the first step of the async work, passing the async_op_ref_for_callback as the context
+                if (handle->ll_execute_async(handle->ll_handle, complete_in_ms, &ll_async_op, ml_async_op_module_with_retries_on_ll_complete_do_retry, (void*)async_op_ref_for_callback) != 0)
+                {
+                    LogError("ll_execute_async for lower module failed");
+                    result = MU_FAILURE;
+                }
+                else
+                {
+                    // 8. Take a lock to synchronize with the possibility of the callback already being called (or any subsequent retries)
+                    srw_lock_ll_acquire_exclusive(&async_op_context->ll_async_op_lock);
+                    {
+                        // 9. Make sure we only store the latest ll_async_op by synchronizing on ll_async_op_epoch
+                        //    This handles the case where the callback for step 1 has already been called and now the ll_async_op in this scope is complete and we don't need it
+                        if (interlocked_compare_exchange(&async_op_context->ll_async_op_epoch, 1, 0) == 0)
+                        {
+                            // 10. Store the ll_async_op in the context on success so that it can be canceled
+                            //     This must happen before the next line so that the ll_async_op is set before cancel is called
+                            THANDLE_MOVE(ASYNC_OP)(&async_op_context->ll_async_op, &ll_async_op);
+                        }
+                        else
+                        {
+                            THANDLE_ASSIGN(ASYNC_OP)(&ll_async_op, NULL);
+                        }
+                        srw_lock_ll_release_exclusive(&async_op_context->ll_async_op_lock);
+                    }
+
+                    // 11. Provide the async_op to the caller
+                    //    After this (but before returning) it is possible that the caller may call cancel because they may have access to the async_op_out pointer in another thread
+                    THANDLE_MOVE(ASYNC_OP)(async_op_out, &async_op);
```

Comment:
Hmmm, but that is error in the caller module.
The caller module should never attempt to use an THANDLE(ASYNC_OP) before we return 0.

That is because there is no way for the caller module to know that 0 will be returned. It seems like the pattern in the caller module is wrong, passing directly some THANDLE(ASYNC_OP)* to be filled in, but not actually checking whether the call succeeded.

I understand how the MOVE here might help, but in all fairness it hides a problem in the caller.